File: src/main/java/br/com/g1bet/security/JWTAuthentication.java
package br.com.g1bet.security;

import br.com.g1bet.model.Usuario;
import com.auth0.jwt.JWT;
import com.auth0.jwt.algorithms.Algorithm;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.http.HttpHeaders;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.AuthenticationException;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import javax.servlet.FilterChain;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;
import java.util.Date;

public class JWTAuthentication extends UsernamePasswordAuthenticationFilter {

    private final AuthenticationManager authenticationManager;

    private static final long EXPIRATION = 600000;

    private static final String senha = "APOSTAS G1BET";

    public JWTAuthentication(AuthenticationManager authenticationManager) {
        this.authenticationManager = authenticationManager;
    }

    @Override
    public Authentication attemptAuthentication(HttpServletRequest request, HttpServletResponse response) throws AuthenticationException {
        // busca as credenciais digitadas pelo usuário
        Usuario usuario = null;
        try {
            usuario = new ObjectMapper().readValue(request.getInputStream(), Usuario.class);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }

        UsernamePasswordAuthenticationToken authResult = new UsernamePasswordAuthenticationToken(
                usuario.getEmail(),
                usuario.getSenha()
        );

        return authenticationManager.authenticate(authResult);
    }

    @Override
    protected void successfulAuthentication(HttpServletRequest request, HttpServletResponse response, FilterChain chain, Authentication authResult) {
        String username = ((UserDetails) authResult.getPrincipal()).getUsername();

        String token = JWT.create()
                .withSubject(username)
                .withExpiresAt(new Date(System.currentTimeMillis() + EXPIRATION))
                .sign(Algorithm.HMAC512(senha.getBytes()));

        String headerBody = username + " " + token;

        response.setHeader(HttpHeaders.AUTHORIZATION , headerBody);
    }
}

File: src/main/java/br/com/g1bet/security/JWTAuthentication.kt
package br.com.g1bet.security

import br.com.g1bet.model.Usuario
import com.auth0.jwt.JWT
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.http.HttpHeaders
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.config.Elements.JWT
import org.springframework.security.core.Authentication
import org.springframework.security.core.AuthenticationException
import org.springframework.security.core.userdetails.UserDetails
import java.io.IOException
import java.util.*
import javax.servlet.FilterChain
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

class JWTAuthentication(authenticationManager: AuthenticationManager) : UsernamePasswordAuthenticationFilter() {
    private val authenticationManager: AuthenticationManager

    init {
        this.authenticationManager = authenticationManager
    }

    @Throws(AuthenticationException::class)
    fun attemptAuthentication(request: HttpServletRequest, response: HttpServletResponse?): Authentication {
        // busca as credenciais digitadas pelo usuário
        var usuario: Usuario? = null
        usuario = try {
            ObjectMapper().readValue(request.getInputStream(), Usuario::class.java)
        } catch (e: IOException) {
            throw RuntimeException(e)
        }
        val authResult = UsernamePasswordAuthenticationToken(
                usuario.getEmail(),
                usuario.getSenha()
        )
        return authenticationManager.authenticate(authResult)
    }

    protected fun successfulAuthentication(request: HttpServletRequest?, response: HttpServletResponse, chain: FilterChain?, authResult: Authentication) {
        val username: String = (authResult.getPrincipal() as UserDetails).getUsername()
        val token: String = JWT.create()
                .withSubject(username)
                .withExpiresAt(Date(System.currentTimeMillis() + EXPIRATION))
                .sign(Algorithm.HMAC512(senha.toByteArray()))
        val headerBody = "$username $token"
        response.setHeader(HttpHeaders.AUTHORIZATION, headerBody)
    }

    companion object {
        private const val EXPIRATION: Long = 600000
        private const val senha = "APOSTAS G1BET"
    }
}

